src/cocaine-app/infrastructure.py

```diff
@@ -167,158 +166,244 @@ def _sync_state(self):
-                        # add group to state only if checks succeeded
-                        self.state[g.group_id] = group_state
-
-                    cur_group_state = (group_state['nodes'] and
-                                       group_state['nodes'][-1]
-                                       or {'set': []})
-
-                    state_nodes = tuple(tuple(nbs)
-                                        for nbs in cur_group_state['set'])
-                    state_nodes_set = set(state_nodes)
-
-                    # extended storage nodes set which includes newly seen nodes,
-                    # do not discard lost nodes
-                    # also filter out old history nodes
-                    ext_storage_nodes = (storage_nodes + tuple(
-                        nb for nb in state_nodes
-                            if nb not in storage_nodes and len(nb) != 2))
-
-                    if set(ext_storage_nodes) != state_nodes_set:
-                        logger.info('Group %d info does not match,'
-                                     'last state: %s, current state: %s' %
-                                     (g.group_id, state_nodes, ext_storage_nodes))
-                        new_nodes = ext_storage_nodes
-
-                    cur_couple_state = (group_state['couples'] and
-                                        group_state['couples'][-1]
-                                        or {'couple': tuple()})
-                    state_couple = cur_couple_state['couple']
-
-                    if storage_couple and set(state_couple) != set(storage_couple):
-                        logger.info('Group %d couple does not match,'
-                                     'last state: %s, current state: %s' %
-                                     (g.group_id, state_couple, storage_couple))
-                        new_couple = storage_couple
-
-                    if new_nodes or new_couple:
-                        try:
-                            self._update_group(g.group_id, new_nodes, new_couple)
-                        except Exception as e:
-                            logger.error('Failed to update infrastructure state for group %s: %s\n%s' %
-                                (g.group_id, e, traceback.format_exc()))
-                            pass
+                    new_couple_record = self._new_group_couple_record(
+                        g, group_history
+                    )
+
+                    if new_node_backends_set_record or new_couple_record:
+                        self._update_group(
+                            group_history,
+                            new_node_backends_set_record,
+                            new_couple_record
+                        )
 
+                except CacheUpstreamError as e:
+                    logger.error('Failed to update history for group {}: {}'.format(
+                        g.group_id, e
+                    ))
+                    failed_groups.add(g)
```

Comment:
logger.exception provides a full traceback, this should happen only in major exception cases. Here CacheUpstreamError is minor case, just a notion in a log will be enough.